Add support for Go 1.16

nacx opened the pull request:
Adds support for Go 1.16 by getting rid of the missing go.sum errors:
```
bingo get github.com/bufbuild/buf/cmd/buf@v0.36.0
Error: get command failed: get: buf.mod: getting github.com/bufbuild/buf/cmd/buf@v0.36.0: install: go: github.com/bufbuild/buf@v0.36.0: missing go.sum entry; to add it:
	go mod download github.com/bufbuild/buf
: exit 1
```

In Go 1.16 the go-modules behavior changed and it no longer updates the `go.mod` and `go.sum` automatically when building or running `go list` (https://blog.golang.org/go116-module-changes).
To overcome this, in Go >= 1.16 we run `go mod downlaod` to make sure it populates what's missing so that `go list` and `go build` succeed.

/cc @codefromthecrypt

codefromthecrypt commented on get_e2e_test.go:
```diff
@@ -692,6 +721,16 @@ func TestGet(t *testing.T) {
 					{
 						name: "Use -u to upgrade thanos package",
 						do: func(t *testing.T) {
+							// In Go 1.16, to get does not give hints about the versions it finds:
```
```suggestion
							// In Go 1.16, 'go get' does not give hints about the versions it finds:
```

codefromthecrypt commented on get_e2e_test.go:
```diff
@@ -670,6 +689,16 @@ func TestGet(t *testing.T) {
+							//             github.com/cortexproject/cortex/pkg/util imports
+							//             google.golang.org/grpc/naming: cannot find module providing package google.golang.org/grpc/naming
+							if !goVersion.LessThan(version.Go116) {
+								t.Skip("skipping pre-modules test in Go >= 1.16; he output of 'go get' does not give hints on the resolved versions")
```
```suggestion
								t.Skip("skipping pre-modules test in Go >= 1.16; the output of 'go get' does not give hints on the resolved versions")
```

codefromthecrypt commented on get_e2e_test.go:
```diff
@@ -713,6 +752,16 @@ func TestGet(t *testing.T) {
 					{
 						name: "Use -u=patch to upgrade thanos package",
 						do: func(t *testing.T) {
+							// In Go 1.16, to get does not give hints about the versions it finds:
```
```suggestion
							// In Go 1.16, 'go get' does not give hints about the versions it finds:
```

codefromthecrypt commented on get_e2e_test.go:
```diff
@@ -713,6 +752,16 @@ func TestGet(t *testing.T) {
+							//             github.com/cortexproject/cortex/pkg/util imports
+							//             google.golang.org/grpc/naming: cannot find module providing package google.golang.org/grpc/naming
+							if !goVersion.LessThan(version.Go116) {
+								t.Skip("skipping pre-modules test in Go >= 1.16; he output of 'go get' does not give hints on the resolved versions")
```
```suggestion
								t.Skip("skipping pre-modules test in Go >= 1.16; the output of 'go get' does not give hints on the resolved versions")
```

codefromthecrypt commented on get_e2e_test.go:
```diff
@@ -670,6 +689,16 @@ func TestGet(t *testing.T) {
 					{
 						name: "Get tricky case with replace (thanos)",
 						do: func(t *testing.T) {
+							// In Go 1.16, to get does not give hints about the versions it finds:
```
```suggestion
							// In Go 1.16, 'go get' does not give hints about the versions it finds:
```

codefromthecrypt commented on get.go:
```diff
@@ -580,6 +581,14 @@ func install(runnable runner.Runnable, name string, link bool, pkg *bingo.Packag
 	}
 
+	// Run go mod download to make sure a proper go.sum file is populated for the module.
+	// Otherwise the next `go list` will fail in Go 1.16
```
```suggestion
	// Otherwise, the next `go list` will fail in Go 1.16
```

codefromthecrypt commented on get.go:
```diff
@@ -580,6 +581,14 @@ func install(runnable runner.Runnable, name string, link bool, pkg *bingo.Packag
 		return errors.Wrap(err, pkg.String())
 	}
 
+	// Run go mod download to make sure a proper go.sum file is populated for the module.
```
```suggestion
	// Run `go mod download` to make sure a proper go.sum file is populated for the module.
```

codefromthecrypt commented on get_e2e_test.go:
```diff
@@ -834,6 +902,15 @@ func TestGet(t *testing.T) {
 
 							// Get all binaries by doing native go build.
 							if isGoProject {
+								// In Go 1.16 we need to explicitly call go mod download to get the go.sum file created. Eagerly call them before go build.
```
```suggestion
								// In Go 1.16, we need to explicitly call 'go mod download' to get the go.sum file created: Eagerly call them before go build.
```

codefromthecrypt commented on get_e2e_test.go:
```diff
@@ -848,6 +925,15 @@ func TestGet(t *testing.T) {
 								testutil.Ok(t, err)
 								testutil.Equals(t, "module_with_replace.buildable 2.7\n", g.ExecOutput(t, p.root, filepath.Join(g.gobin, "wr_buildable-v0.0.0-20210109165512-ccbd4039b94a")))
 							} else {
+								// In Go 1.16 we need to explicitly call go mod download to get the go.sum file created. Eagerly call them before go build.
```
```suggestion
								// In Go 1.16, we need to explicitly call 'go mod download' to get the go.sum file created: Eagerly call them before go build.
```

codefromthecrypt reviewed (commented):
started with a typo pass

bwplotka commented on get_e2e_test.go:
```diff
@@ -670,6 +689,16 @@ func TestGet(t *testing.T) {
+							//             github.com/cortexproject/cortex/pkg/util imports
+							//             google.golang.org/grpc/naming: cannot find module providing package google.golang.org/grpc/naming
+							if !goVersion.LessThan(version.Go116) {
+								t.Skip("skipping pre-modules test in Go >= 1.16; the output of 'go get' does not give hints on the resolved versions")
```
I don't think that is a solution, ideally we fix this too, right? ( :

nacx replied:
Sure, if the version info can be found yes, let's get this fixed as well

bwplotka replied:
Will look around

bwplotka reviewed (commented):
Nice! Thanks for help. Do you mind if I  propose a PR to your PR or send commit? I am playing with it to fix the skipped tests 🤗

bwplotka commented on get_e2e_test.go:
```diff
@@ -603,6 +609,10 @@ func TestGet(t *testing.T) {
 						// Regression test against https://github.com/bwplotka/bingo/issues/65.
 						name: "get tool with capital letters in name (pre modules)",
 						do: func(t *testing.T) {
+							if !goVersion.LessThan(version.Go116) {
```
Do we know why this happens? 🤔 

````
    get_e2e_test.go:612: get_e2e_utils_test.go:229:
        
         unexpected error: error while running command "/tmp/bingo-tmpgoenv193949100/bingo get github.com/githubnemo/CompileDaemon@v1.2.1"; out: Error: get command failed: get: compiledaemon.mod: getting github.com/githubnemo/CompileDaemon@v1.2.1: install: build versioned: ../../gopath/pkg/mod/github.com/githubnemo/!compile!daemon@v1.2.1/daemon.go:79:2: no required module provides package github.com/fatih/color; to add it:
        	go get github.com/fatih/color
        ../../gopath/pkg/mod/github.com/githubnemo/!compile!daemon@v1.2.1/daemon.go:80:2: no required module provides package github.com/fsnotify/fsnotify; to add it:
        	go get github.com/fsnotify/fsnotify
        : exit 1
        ; err: exit status 1
        
            --- FAIL: TestGet/empty_project_with_advanced_cases/isGoProject=false/get_tool_with_capital_letters_in_name_(pre_modules) (5.90s) 
    ```

bwplotka replied:
Actually it's more complex:

```
  
         unexpected error: error while running command "/tmp/bingo-tmpgoenv281836461/bingo get -v github.com/githubnemo/CompileDaemon@v1.2.1"; out: getting target github.com/githubnemo/CompileDaemon@v1.2.1 (module  )
        exec 'go mod init -modfile=/tmp/bingo-tmpgoenv281836461/testproject/.bingo/compiledaemon-e.tmp.mod _'
        exec 'go get -modfile=/tmp/bingo-tmpgoenv281836461/testproject/.bingo/compiledaemon-e.tmp.mod -d github.com/githubnemo/CompileDaemon@v1.2.1'
        exec 'go env GOPATH'
        exec 'go mod init -modfile=/tmp/bingo-tmpgoenv281836461/testproject/.bingo/compiledaemon.tmp.mod _'
        exec 'go mod download -x -modfile=/tmp/bingo-tmpgoenv281836461/testproject/.bingo/compiledaemon.tmp.mod'
        exec 'go list -modfile=/tmp/bingo-tmpgoenv281836461/testproject/.bingo/compiledaemon.tmp.mod -f={{.Name}} github.com/githubnemo/CompileDaemon'
        Error: install: build versioned: ../../gopath/pkg/mod/github.com/githubnemo/!compile!daemon@v1.2.1/daemon.go:79:2: no required module provides package github.com/fatih/color; to add it:
        	go get github.com/fatih/color
        ../../gopath/pkg/mod/github.com/githubnemo/!compile!daemon@v1.2.1/daemon.go:80:2: no required module provides package github.com/fsnotify/fsnotify; to add it:
        	go get github.com/fsnotify/fsnotify
        : error while running command 'go build -modfile=/tmp/bingo-tmpgoenv281836461/testproject/.bingo/compiledaemon.tmp.mod -o=/tmp/bingo-tmpgoenv281836461/bin/compiledaemon-v1.2.1 github.com/githubnemo/CompileDaemon'; err: exit status 1
        compiledaemon.mod: getting github.com/githubnemo/CompileDaemon@v1.2.1
        main.get
```

bwplotka replied:
Found root cause: 
go mod download does not resolve anything, so in go.mod we have only direct dep. Only `go get` adds `go.sum` files for all deps too.

nacx replied:
I also found that the current code does not work well in some cases, such as commits specified directly. This one is failing as well:
```
$ bingo -v get github.com/envoyproxy/protoc-gen-validate@d6164de4910977d3c3c8dbd9299b5064ea9e7c2b

getting target github.com/envoyproxy/protoc-gen-validate@d6164de4910977d3c3c8dbd9299b5064ea9e7c2b (module  )
exec 'go mod init -modfile=/Users/nacx/src/go/src/github.com/tetrateio/ngac/.bingo/protoc-gen-validate-e.tmp.mod _'
exec 'go get -modfile=/Users/nacx/src/go/src/github.com/tetrateio/ngac/.bingo/protoc-gen-validate-e.tmp.mod -d github.com/envoyproxy/protoc-gen-validate@d6164de4910977d3c3c8dbd9299b5064ea9e7c2b'
exec 'go env GOPATH'
exec 'go mod init -modfile=/Users/nacx/src/go/src/github.com/tetrateio/ngac/.bingo/protoc-gen-validate.tmp.mod _'
exec 'go mod download -x -modfile=/Users/nacx/src/go/src/github.com/tetrateio/ngac/.bingo/protoc-gen-validate.tmp.mod'
exec 'go list -modfile=/Users/nacx/src/go/src/github.com/tetrateio/ngac/.bingo/protoc-gen-validate.tmp.mod -f={{.Name}} github.com/envoyproxy/protoc-gen-validate'
Error: install: build versioned: ../../../../../pkg/mod/github.com/envoyproxy/protoc-gen-validate@v0.0.15-0.20190405222122-d6164de49109/gogoproto/gogo.pb.go:15:8: no required module provides package github.com/golang/protobuf/proto; to add it:
	go get github.com/golang/protobuf/proto
../../../../../pkg/mod/github.com/envoyproxy/protoc-gen-validate@v0.0.15-0.20190405222122-d6164de49109/gogoproto/gogo.pb.go:18:8: no required module provides package github.com/golang/protobuf/protoc-gen-go/descriptor; to add it:
	go get github.com/golang/protobuf/protoc-gen-go/descriptor
../../../../../pkg/mod/github.com/envoyproxy/protoc-gen-validate@v0.0.15-0.20190405222122-d6164de49109/templates/cc/register.go:10:2: no required module provides package github.com/golang/protobuf/ptypes; to add it:
	go get github.com/golang/protobuf/ptypes
../../../../../pkg/mod/github.com/envoyproxy/protoc-gen-validate@v0.0.15-0.20190405222122-d6164de49109/validate/validate.pb.go:10:8: no required module provides package github.com/golang/protobuf/ptypes/duration; to add it:
	go get github.com/golang/protobuf/ptypes/duration
../../../../../pkg/mod/github.com/envoyproxy/protoc-gen-validate@v0.0.15-0.20190405222122-d6164de49109/validate/validate.pb.go:11:8: no required module provides package github.com/golang/protobuf/ptypes/timestamp; to add it:
	go get github.com/golang/protobuf/ptypes/timestamp
../../../../../pkg/mod/github.com/envoyproxy/protoc-gen-validate@v0.0.15-0.20190405222122-d6164de49109/templates/java/register.go:12:2: no required module provides package github.com/iancoleman/strcase; to add it:
	go get github.com/iancoleman/strcase
../../../../../pkg/mod/github.com/envoyproxy/protoc-gen-validate@v0.0.15-0.20190405222122-d6164de49109/templates/shared/context.go:11:2: no required module provides package github.com/lyft/protoc-gen-star; to add it:
	go get github.com/lyft/protoc-gen-star
../../../../../pkg/mod/github.com/envoyproxy/protoc-gen-validate@v0.0.15-0.20190405222122-d6164de49109/templates/cc/register.go:14:2: no required module provides package github.com/lyft/protoc-gen-star/lang/go; to add it:
	go get github.com/lyft/protoc-gen-star/lang/go
: error while running command 'go build -modfile=/Users/nacx/src/go/src/github.com/tetrateio/ngac/.bingo/protoc-gen-validate.tmp.mod -o=/Users/nacx/src/go/bin/protoc-gen-validate-v0.0.15-0.20190405222122-d6164de49109 github.com/envoyproxy/protoc-gen-validate'; err: exit status 1
protoc-gen-validate.mod: getting github.com/envoyproxy/protoc-gen-validate@d6164de4910977d3c3c8dbd9299b5064ea9e7c2b
main.get
	/Users/nacx/src/go/src/github.com/bwplotka/bingo/get.go:296
main.main.func2
	/Users/nacx/src/go/src/github.com/bwplotka/bingo/main.go:156
main.main.func5
	/Users/nacx/src/go/src/github.com/bwplotka/bingo/main.go:247
github.com/oklog/run.(*Group).Run.func1
	/Users/nacx/src/go/pkg/mod/github.com/oklog/run@v1.1.0/group.go:38
runtime.goexit
	/usr/local/opt/go/libexec/src/runtime/asm_amd64.s:1371
get
main.main.func2
	/Users/nacx/src/go/src/github.com/bwplotka/bingo/main.go:157
main.main.func5
	/Users/nacx/src/go/src/github.com/bwplotka/bingo/main.go:247
github.com/oklog/run.(*Group).Run.func1
	/Users/nacx/src/go/pkg/mod/github.com/oklog/run@v1.1.0/group.go:38
runtime.goexit
	/usr/local/opt/go/libexec/src/runtime/asm_amd64.s:1371
get command failed
main.main
	/Users/nacx/src/go/src/github.com/bwplotka/bingo/main.go:255
runtime.main
	/usr/local/opt/go/libexec/src/runtime/proc.go:225
runtime.goexit
	/usr/local/opt/go/libexec/src/runtime/asm_amd64.s:1371
```
We should add that one as an e2e test case as well.

bwplotka replied:
Is that failing after my PR too? (:

nacx replied:
Your commit fixes this :)

nacx commented:
>Do you mind if I propose a PR to your PR or send commit? 

Totally! Whatever works best to get Go 1.16 supported. I'm not the most versed in how Go modules work internally, so the approach in this PR may not be the best. Any help is very welcome!

bwplotka commented:
Added PR https://github.com/nacx/bingo/pull/1 if you want to merge/check. We can then iterate over other errors. @nacx 🤗

bwplotka commented:
Thanks for merging, let's iterate, I will get back to this tomorrow (focusing on book writing today)

nacx commented:
Ok, so now the only test that seems to be failing is the `thanos` one, and the error produced does not contain a hint on the version. Playing a bit manually with some commands, I've found that:

We could try the go mod download as json to resolve the `Version` field:
```bash
$ go mod download -json github.com/envoyproxy/protoc-gen-validate@d6164de4910977d3c3c8dbd9299b5064ea9e7c2b
{
	"Path": "github.com/envoyproxy/protoc-gen-validate",
	"Version": "v0.0.15-0.20190405222122-d6164de49109",
	"Info": "/Users/nacx/src/go/pkg/mod/cache/download/github.com/envoyproxy/protoc-gen-validate/@v/v0.0.15-0.20190405222122-d6164de49109.info",
	"GoMod": "/Users/nacx/src/go/pkg/mod/cache/download/github.com/envoyproxy/protoc-gen-validate/@v/v0.0.15-0.20190405222122-d6164de49109.mod",
	"Zip": "/Users/nacx/src/go/pkg/mod/cache/download/github.com/envoyproxy/protoc-gen-validate/@v/v0.0.15-0.20190405222122-d6164de49109.zip",
	"Dir": "/Users/nacx/src/go/pkg/mod/github.com/envoyproxy/protoc-gen-validate@v0.0.15-0.20190405222122-d6164de49109",
	"Sum": "h1:U1RyeEwqO++6RvIA7bY98AcYIFPQMwrOeWNmw+dvq9w=",
	"GoModSum": "h1:iSmxcyjqTsJpI2R4NaDN7+kN2VEUnK/pcBlmesArF7c="
}
```
For the Thanos one, though, since it is using `/cmd/thanos` and not the root the command fails, but the error still gives a hint on the resolved version:
```bash
 $ go mod download -json  github.com/thanos-io/thanos/cmd/thanos@f85e4003ba51f0592e42c48fdfdf0b800a23ba74
{
	"Path": "github.com/thanos-io/thanos/cmd/thanos",
	"Version": "f85e4003ba51f0592e42c48fdfdf0b800a23ba74",
	"Error": "github.com/thanos-io/thanos/cmd/thanos@v0.0.0-20210108102609-f85e4003ba51: invalid version: missing github.com/thanos-io/thanos/cmd/thanos/go.mod at revision f85e4003ba51"
}
```
Is this something we could use when handling this errors?

bwplotka commented:
Thanks, that sounds like some solution, I can look on it only in few hours.

Ideally we don't need to extra calls that's my only worry 🤗

nacx commented:
I've been playing around with this a bit, and added a commit with an idea. Basically, when the error does not contain any hint about the module version, we can try find it in the local module cache and resolve that version.
I've added a small (a bit dirty) function that implements this to see if it fixes the remaining e2e tests.

bwplotka commented:
Thanks so much, I got out from meetings so looking on this. I wonder if we can do that by default (: Wil push commit in second.

bwplotka commented:
I think the fix might be easier then we though, trying here: https://github.com/bwplotka/bingo/pull/75 (:

bwplotka commented:
Pushed potential "fix".

bwplotka commented:
Nope, not yet done (:

bwplotka commented:
Last standing error is this: https://github.com/golang/go/blob/master/src/cmd/go/internal/modload/query.go#L849

bwplotka commented:
https://github.com/golang/go/commit/06538fa723fc358462c5d7ae385b5b64ac76827b

bwplotka commented:
Wow. it's so unreliable at this point:
```
   unexpected error: error while running command "/tmp/bingo-tmpgoenv444360868/bingo get github.com/thanos-io/thanos/cmd/thanos@f85e4003ba51f0592e42c48fdfdf0b800a23ba74"; out: Error: get command failed: get: thanos.mod: getting github.com/thanos-io/thanos/cmd/thanos@f85e4003ba51f0592e42c48fdfdf0b800a23ba74: go get did not found the package (or none of our regexps matches: go: found github.com/thanos-io/thanos/cmd/thanos in (\S*) (\S*),go: downloading (github.com/thanos-io/thanos/cmd) (\S*),go: downloading (github.com/thanos-io/thanos) (\S*),go: (\S*) upgrade => (\S*)): go get: github.com/thanos-io/thanos@v0.11.0 updating to
```

Time for `go mod --json` solution I suppose.

nacx commented:
Would it be more efficient to walk the local cache as in the latter attempt instead of doing a go mod download? That could save some network calls and be potentially faster? One caveat there is that with that approach, it failed on the `upgrade` test. For the thanos example it resolved the version properly but the upgrade test failed (I still haven't had time to fully investigate that failure).

bwplotka commented:
Yes, that was my worry is that this kind of works, but then we don't know which version "would be" actually chosen by go get, if you have many dirs cached. The only way to know is to... call go mod download again. This is so funky, I wished the internal Go code was not under `internal`.... (:

bwplotka commented:
And I think extra network call is not that bad solution - better than nothing 🙃

bwplotka commented:
If you are worried by network calls, check -x and what go get is doing ((:  (dozen of calls)

nacx commented:
Hmmm Looking at the `thanos` example in the `go mod download`:
```json
{
	"Path": "github.com/thanos-io/thanos/cmd/thanos",
	"Version": "f85e4003ba51f0592e42c48fdfdf0b800a23ba74",
	"Error": "github.com/thanos-io/thanos/cmd/thanos@v0.0.0-20210108102609-f85e4003ba51: invalid version: missing github.com/thanos-io/thanos/cmd/thanos/go.mod at revision f85e4003ba51"
}
```
Note that the error message (and the path) refer to the given input (/cmd/thanos) and **not** to the directory that contains the actual `go.mod` (github.com/thanos-io/thanos). Would this be an issue? Because in the code that walks the cache we could properly detect that, but not in this case...

bwplotka commented:
Yea I  would do the same as I did now: 

```go
	// Since we don't know which part of full path is package, which part is module, start from longest and go until we find one.
		for i := len(strings.Split(target.Path(), "/")); i > 3; i-- {
			modulePath = filepath.Dir(modulePath)

			attempted = append(attempted, fmt.Sprintf(`go: downloading (%v) (\S*)`, modulePath))
			re, err = regexp.Compile(attempted[len(attempted)-1])
			if err != nil {
				return errors.Wrapf(err, "regexp compile %v", attempted[len(attempted)-1])
			}
			if re.MatchString(gerr.Error()) {
				break
			}
		}
		if !re.MatchString(gerr.Error()) {
			attempted = append(attempted, `go: (\S*) upgrade => (\S*)`)
			re = regexp.MustCompile(attempted[len(attempted)-1])
		}
	}
```

Just with go mod download. WDYT?

bwplotka commented:
But then again, this solution is short on one more thing: When you don't specify any version, and if go-get fails then we don't know updated version too sooooo, we need something else. ):

nacx commented:
Thinking out loud now:
With the local cache thing, we should always have a version. If the version is explicitly set in the command line, we should be good.
If not, given that we're creating empty go.mod files with just one line, we can assume that if no version is specified in the command line go get would resolve to the latest one, and we should be able to get that latest version from the cache as well by inspecting: `$(GOMODCACHE)/cache/download/github.com/thanos-io/thanos/@v/list`
Is this something that could work?

bwplotka commented:
I think that is a solid workaround. Sorry for killing your gomodcache try out - you can recover it from https://github.com/bwplotka/bingo/pull/74/commits/4cccf2865bc07fcd0b2e3c62786911fb9e1a95d8 🤗

bwplotka commented:
I think it might not much the go get update logic 1:1 but we can have something very simplified - we could even try to remove `-u` and `-upatch` and just have `-u`  which is very trivial - just get latest

bwplotka commented:
It's 11pm and I have been working from 9am, so time for me to stop - I will fix this tomorrow if you will not do that earlier. Thanks for help!

And btw, we can then remove output matching totally IMO (for all Go versions)

nacx commented:
NP! Let's see if we can make this work :) I have to go AFK now, but I'll give it a try tomorrow if you don't beat me at it (feel free to continue experimenting in this same PR!)

bwplotka commented:
Ok, focusing on it now, let's go.

bwplotka commented:
![](https://media0.giphy.com/media/dIxkmtCuuBQuM9Ux1E/giphy.gif)

nacx commented:
And it's working indeed! Yay!

bwplotka commented:
Thanks for great idea! Releasing 🤗

nacx commented on get.go:
```diff
@@ -391,51 +395,126 @@ func resolvePackage(
+	if lastVersion == "" {
+		return "", errors.New("empty file")
+	}
+	return lastVersion, nil
```
If I'm not wrong, this returns the last version found in the file, which seems to be how the file is laid out. Given that the `semver` library already has functions to compare, is it worth sorting the versions based on semver instead of returning the last one?

bwplotka replied:
I don't think this will work, since there are those custom versions too (with sha and date). But maybe it follows semver (: Then yes.

In practice what I checked this file is sorted, but you are right it would be better to ensure this just in case as this behavior is not stated anywhere. Let's do in separate PR
